ipaserver/plugins/cert.py

```diff
@@ -1046,6 +1054,84 @@ def _principal_name_matches_principal(name, principal_obj):
     return principal in principal_obj.get('krbprincipalname', [])
 
 
+def _validate_san_ips(generalnames):
+    """
+    Raise a ValidationError if the subjectAltName in a CSR includes
+    any IP address(es) that do not match a DNS name in the SAN.
+
+    """
+    san_dns_ips = set()
+    for name in generalnames:
+        if isinstance(name, cryptography.x509.general_name.DNSName):
+            san_dns_ips.update(_san_dnsname_ips(name.value))
+    for name in generalnames:
+        if isinstance(name, cryptography.x509.general_name.IPAddress):
+            if unicode(name.value) not in san_dns_ips:
+                raise errors.ValidationError(
+                    name='csr',
+                    error=_(
+                        "IP address in subjectAltName (%s) does not "
+                        "match any DNS name"
+                    ) % name.value
+                )
+
+
+def _san_dnsname_ips(dnsname, dnsname_is_cname=False):
+    """
+    Returns a set of IP addresses, managed by this IPa instance,
+    that correspond to the DNS name (from the subjectAltName).
+
+    """
+    fqdn = dnsutil.DNSName(dnsname).make_absolute()
+    if fqdn.__len__() < 4:
+        logger.debug("Skipping IPs for %s: hostname too short" % dnsname)
+        return ()
+    zone = dnsutil.DNSName(resolver.zone_for_name(fqdn))
```

Comment:
Instead of approaching it this way, where we potentially forward to external DNS servers that
may not respond promptly, perhaps we can perform `ipa dnszone-find` or direct LDAP searches to see
whether the IPA DNS serves a zone that includes the given name.

I propose an ldapsearch, otherwise we may have to repeat the `dnszone-find`, removing the leftmost domain component, until we either find a match or hit the root zone.

This ldapsearch can also take the place of the `dnsrecord-show` command.

Let me play around with the IPA DNS for a bit (I am not familiar with this subsystem) and see how we can do it.

Reply:
We could do that, but it's more work.  :-)

Seriously, this is getting into optimizing the error path.  We'll only hit DNS timeout issues in the case where the CSR SAN includes a hostname or IP that isn't managed by this IPA instance.  Is it worth spending effort on this case?